R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,209 @@
+    "--output-lib" "${CMAKE_CURRENT_BINARY_DIR}/R.lib")
+endfunction(create_rlib_for_msvc)
+
+# detection for OSX
+if(APPLE)
+
+  find_library(LIBR_LIBRARIES R)
+
+  if(LIBR_LIBRARIES MATCHES ".*\\.framework")
+    set(LIBR_HOME "${LIBR_LIBRARIES}/Resources" CACHE PATH "R home directory")
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include" CACHE PATH "R include directory")
+    set(LIBR_EXECUTABLE "${LIBR_HOME}/R" CACHE PATH "R executable")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/lib" CACHE PATH "R lib directory")
+  else()
+    get_filename_component(_LIBR_LIBRARIES "${LIBR_LIBRARIES}" REALPATH)
+    get_filename_component(_LIBR_LIBRARIES_DIR "${_LIBR_LIBRARIES}" DIRECTORY)
+    set(LIBR_EXECUTABLE "${_LIBR_LIBRARIES_DIR}/../bin/R")
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--vanilla" "-e" "cat(R.home())"
+      OUTPUT_VARIABLE LIBR_HOME
+    )
+    set(LIBR_HOME ${LIBR_HOME} CACHE PATH "R home directory")
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include" CACHE PATH "R include directory")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/lib" CACHE PATH "R lib directory")
+  endif()
+
+# detection for UNIX & Win32
+else()
+
+  # attempt to find R executable
+  if(NOT LIBR_EXECUTABLE)
+
+    # CRAN may run RD CMD CHECK instead of R CMD CHECK,
+    # which can lead to this infamous error:
+    # 'R' should not be used without a path -- see par. 1.6 of the manual
+    find_program(
+      LIBR_EXECUTABLE
+      NO_DEFAULT_PATH
+      HINTS "${CMAKE_CURRENT_BINARY_DIR}" "/usr/bin" "/usr/lib/"
+      NAMES R R.exe
+    )
+
+    if(LIBR_EXECUTABLE MATCHES ".*lightgbm\\.Rcheck.*")
+      message(FATAL_ERROR "Hit this block. '${LIBR_EXECUTABLE}'")
+    endif()
+  endif()
+
+  if(UNIX)
+
+    if(NOT LIBR_EXECUTABLE)
+      message(FATAL_ERROR "Unable to locate R executable.\
+        \nEither add its location to PATH or provide it through the LIBR_EXECUTABLE cmake variable")
+    endif()
+
+    # ask R for the home path
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--vanilla" "-e" "cat(R.home())"
+      OUTPUT_VARIABLE LIBR_HOME
+    )
+
+    # ask R for the include dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('include'))"
+      OUTPUT_VARIABLE LIBR_INCLUDE_DIRS
+    )
+
+    # ask R for the lib dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('lib'))"
+      OUTPUT_VARIABLE LIBR_LIB_DIR
+    )
+
+  # Windows
+  else()
+
+    # ask R for R_HOME
+    if(LIBR_EXECUTABLE)
+      execute_process(
+        COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(normalizePath(R.home(),winslash='/'))"
+        OUTPUT_VARIABLE LIBR_HOME
+      )
+    endif()
+
+    # if R executable not available, query R_HOME path from registry
+    if(NOT LIBR_HOME)
+
+      # Windows users might want to change this to their R version:
+      if(NOT R_VERSION)
+        set(R_VERSION "3.4.1")
```

Comment:
oh nice find!  Ok let me try with that.

Reply:
ok!  Let me try that

Reply:
Actually....since we're already using `install.libs.R`, R had to have already been found for this code to even be invoked. I'm going to try just using R's built-in `R.Version()` and passing it in as a cmake variable, e.g. `cmake .. -DCMAKE_R_VERSION=3.6.1`

Reply:
^  I tried this (just passing it in from R) in https://github.com/microsoft/LightGBM/pull/2901/commits/9a337d4f4d4004346d7d8d64ced4a0b719fc07ae and it seemed to work locally. Let's see if CI agrees with me.

I think that is a better solution than consulting the registry, since the things you linked mention that creating those registry entries is optional. I think the average R on Windows user will probably just have done an install with all the default settings.